### PR TITLE
Fix race conditions in poll and condvar

### PR DESCRIPTION
**Pull request by ambroise-arm**
In the poll code, `signal_poll_event` is called by 2 functions: `z_impl_k_poll_signal_raise` and `z_handle_obj_poll_events`. The first one correctly takes the poll lock. Fix the second one not taking the lock.

In the sockets code, fix the mutex of the condition variable getting released before signaling for this same condition variable.

Both of those created race conditions which especially appeared on SMP-enabled configurations. The symptom was similar for both of them: a thread pending, waiting on an event, but never getting scheduled again even though the event happened.

### CODE REVIEW

**Review by jukkar — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Review by povergoing — APPROVED** [review]
I am just wondering why the naming of `z_handle_obj_poll_events` looks so different from others in [kernel/poll.c](https://github.com/zephyrproject-rtos/zephyr/pull/62691/files#diff-06e616b4678c2bb4ee4b383f66a36821aa1940dc7bf20b8a3f68b1573bb02875).

**Comment by nashif** [comment]
@peter-mitsis  @andyross  please review.

**Review by peter-mitsis — APPROVED** [review]

File: kernel/poll.c [hunk 1]
```diff
@@ -470,11 +471,14 @@ static int signal_poll_event(struct k_poll_event *event, uint32_t state)
 void z_handle_obj_poll_events(sys_dlist_t *events, uint32_t state)
 {
 	struct k_poll_event *poll_event;
+	k_spinlock_key_t key = k_spin_lock(&lock);
```
Comment (andyross) [inline kernel/poll.c:474]: Definitely looks like a lock was missing here, but I do worry about the possibility of a deadlock underneath signal_poll_event, which calls into a bunch of subsystems.  Are we 100% sure those can't turn around and call back into something that wants the poll lock?

**Review by andyross — COMMENTED** [review]
One question.  If you're already done the analysis and know it's safe, feel free to treat your "yes" as a +1 from me.

**Review by andyross — APPROVED** [review]
I'm sold.

**Review by nashif — APPROVED** [review]

**Pull request merged**